### PR TITLE
Async-streams: Call AsyncIteratorMethodBuilder.Complete

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/39321

We need to call `AsyncIteratorMethodBuilder.Complete` to dispose of some resources it holds. We can do that when bringing the state machine to a finalized state (`-2`), that is when we'd call `AsyncTaskMethodBuilder.SetResult` in a plain async method.
But a completed instance cannot be re-used, so we also need to reset it in `GetAsyncEnumerator`.

For more context, here's the code we [currently produce](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWDkQKyZiwGZkxEBhRAb1kTuRJAA5EBJAQQGcBPAOwGMAorwCuAWwCmAJwCGAIwA2EgDwBLXgBcAfIgCyACgCU1WvTMgAnMgBsAOgCaqiQoAmRgmfMQIyAOyJ4D3oAX1hgoA===).

Here's a sample showing an overview of the change:

``` csharp
	private sealed class <M>d__0 : IAsyncEnumerable<int>, IAsyncEnumerator<int>, IAsyncDisposable, IValueTaskSource<bool>, IValueTaskSource, IAsyncStateMachine
	{
		public int <>1__state;
		public AsyncIteratorMethodBuilder <>t__builder;
		public ManualResetValueTaskSourceCore<bool> <>v__promiseOfValueOrEnd;
		private int <>2__current;
		private bool <>w__disposeMode;
		private int <>l__initialThreadId;
		private TaskAwaiter <>u__1;

		[DebuggerHidden]
		public <M>d__0(int <>1__state)
		{
			this.<>1__state = <>1__state;
			<>l__initialThreadId = Environment.CurrentManagedThreadId;
                        <>t__builder = AsyncIteratorMethodBuilder.Create();
		}

		private void MoveNext()
		{
			int num = <>1__state;
			try
			{
                            ...
			}
			catch (Exception exception)
			{
				<>1__state = -2;
				<>v__promiseOfValueOrEnd.SetException(exception);
				<>t__builder.Complete(); // added
				return;
			}
			<>1__state = -2;
			<>v__promiseOfValueOrEnd.SetResult(result: false);
			<>t__builder.Complete(); // added
			return;
			IL_011f:
			<>v__promiseOfValueOrEnd.SetResult(result: true);
		}

		[DebuggerHidden]
		IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(CancellationToken token)
		{
			if (<>1__state == -2 && <>l__initialThreadId == Environment.CurrentManagedThreadId)
			{
				<>1__state = -3;
				<>t__builder = AsyncIteratorMethodBuilder.Create(); // added
				<>w__disposeMode = false;
				return this;
			}
			return new <M>d__0(-3);
		}

            ...
	}
```

### CODE REVIEW

**Comment by jcouv** [comment]
Rebased to resolve conflict

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Please review. I'd like to get this fix in before tomorrow's snap. Thanks

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs [hunk 1]
```diff
@@ -2666,32 +2685,40 @@ .maxstack  3
...
+  IL_0126:  nop
+  IL_0127:  ldarg.0
+  IL_0128:  ldflda     ""System.Runtime.CompilerServices.AsyncIteratorMethodBuilder C.<M>d__0.<>t__builder""
+  IL_012d:  call       ""void System.Runtime.CompilerServices.AsyncIteratorMethodBuilder.Complete()""
+  IL_0132:  nop
+  IL_0133:  ret
+  IL_0134:  ldarg.0
+  IL_0135:  ldflda     ""System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
+  IL_013a:  ldc.i4.1
+  IL_013b:  call       ""void System.Threading.Tasks.Sources.ManualResetValueTaskSourceCore<bool>.SetResult(bool)""
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: Why don't we need to call `AsyncIteratorMethodBuilder.Complete` here?

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs [hunk 2]
```diff
@@ -102,6 +104,19 @@ BoundExpressionStatement generateSetResultOnPromise(bool result)
             }
         }
 
+        BoundExpressionStatement GenerateCompleteOnBuilder()
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs]: >B [](start = 8, length = 1)

`private`

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs [hunk 3]
```diff
@@ -102,6 +104,19 @@ BoundExpressionStatement generateSetResultOnPromise(bool result)
...
+        {
+            // Produce:
+            // this.builder.Complete();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs]: >this.builder.Complete(); [](start = 15, length = 24)

It looks like the code below is `builder.SetResult(value)`.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs [hunk 4]
```diff
@@ -102,6 +104,19 @@ BoundExpressionStatement generateSetResultOnPromise(bool result)
...
+            return F.ExpressionStatement(
+                F.Call(
+                    F.Field(F.This(), _asyncMethodBuilderField),
+                        _asyncMethodBuilderMemberCollection.SetResult, // AsyncIteratorMethodBuilder.Complete is the corresponding method to AsyncTaskMethodBuilder.SetResult
+                        _method.IsGenericTaskReturningAsync(F.Compilation)
+                            ? ImmutableArray.Create<BoundExpression>(F.Local(_exprRetValue))
+                            : ImmutableArray<BoundExpression>.Empty));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs]: Is there an argument?

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs [hunk 5]
```diff
@@ -632,10 +632,31 @@ private void GenerateIAsyncEnumerableImplementation_GetAsyncEnumerator()
...
+                // this.disposeMode = false;
+
+                builder.Add(
+                    // this.state = {initialState};
+                    F.Assignment(F.Field(F.This(), stateField), F.Literal(initialState)));
+
+                // this.builder = System.Runtime.CompilerServices.AsyncVoidMethodBuilder.Create();
+                AsyncMethodBuilderMemberCollection methodScopeAsyncMethodBuilderMemberCollection;
+                bool found = AsyncMethodBuilderMemberCollection.TryCreate(F, method, typeMap: null, out methodScopeAsyncMethodBuilderMemberCollection);
+                Debug.Assert(found);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs]: The base class has an `_asyncMethodBuilderMemberCollection` field. Consider using that here rather than calling `TryCreate()` again which does extra work (and might produce duplicate warnings?). Same comment for `GenerateConstructor()` above.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs]: You're correct that I re-used this call from a call to SetResult. I should have removed the argument (`SetResult` never has an argument). Thanks!

---
In reply to: [337668623](https://github.com/dotnet/roslyn/pull/39436#discussion_r337668623) [](ancestors = 337668623)

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs [hunk 6]
```diff
@@ -133,7 +133,7 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
                     createBuilderMethod: createBuilderMethod,
                     taskProperty: null,
                     setException: null, // unused
-                    setResult: WellKnownMember.System_Runtime_CompilerServices_AsyncIteratorMethodBuilder__Complete,
+                    setResult: WellKnownMember.System_Runtime_CompilerServices_AsyncIteratorMethodBuilder__Complete, // AsyncIteratorMethodBuilder.Complete is the corresponding method to AsyncTaskMethodBuilder.SetResult
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs]: I find this kind of a confusing pattern, but I'm not sure how I would change it

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: The tail of the `MoveNext()` method is like this:
``` 
            // ... _exprReturnLabel: ...
            // ... this.state = FinishedState; ...
            // if (this.combinedTokens != null) { this.combinedTokens.Dispose(); this.combinedTokens = null; } // for enumerables only
            // this.promiseOfValueOrEnd.SetResult(false);
            // this.builder.Complete();
            // return;

            // _exprReturnLabelTrue:
            // this.promiseOfValueOrEnd.SetResult(true);

            // ... _exitLabel: ...
            // ... return; ...
```

It has different ways of returning from `MoveNext()` (in non-exceptional flow): 
- one for signaling that we have no values left: we set the finished state, dispose our `CancellationTokenSource`, record that we're done with `SetResult(false)` and dispose the builder with `Complete()`.
- the other for returning a value in the middle of the stream: we have already save that value in `Current`, we just signal with `SetResult(true)`. We don't need to dispose any resources, since we expect to process more values in the stream.

---
In reply to: [337657566](https://github.com/dotnet/roslyn/pull/39436#discussion_r337657566) [](ancestors = 337657566)

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: Also it helps to look at the example I linked to in description showing current tail of `MoveNext()`. Only two of the exit paths need to dispose.
```
            }
            catch (Exception exception)
            {
                <>1__state = -2;
                <>v__promiseOfValueOrEnd.SetException(exception);
                return;
            }
            <>1__state = -2;
            <>v__promiseOfValueOrEnd.SetResult(false);
            return;
            IL_00ed:
            <>v__promiseOfValueOrEnd.SetResult(true);
        }
```

---
In reply to: [337681134](https://github.com/dotnet/roslyn/pull/39436#discussion_r337681134) [](ancestors = 337681134,337657566)

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs (hunk 6 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs]: The BCL initially had AsyncIteratorMethodBuilder.SetResult (like other builders), but then changed the name since there is no value/result. Still is it the spiritual equivalent... ;-)

---
In reply to: [337678803](https://github.com/dotnet/roslyn/pull/39436#discussion_r337678803) [](ancestors = 337678803)

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs [hunk 7]
```diff
@@ -164,23 +164,25 @@ protected override void GenerateConstructor()
...
-                        F.InstanceField(_builderField),
-                        F.StaticCall(
-                            null,
-                            methodScopeAsyncMethodBuilderMemberCollection.CreateBuilder)));
+                // this.builder = System.Runtime.CompilerServices.AsyncIteratorMethodBuilder.Create();
+                bodyBuilder.Add(GenerateCreateAndAssignBuilder());
 
                 bodyBuilder.Add(F.Return());
                 F.CloseMethod(F.Block(bodyBuilder.ToImmutableAndFree()));
                 bodyBuilder = null;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs]: >bodyBuilder = null; [](start = 16, length = 19)

Why assign `null`?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs]: This was from a previous code review if I recall correctly. I don't care much, but it does prevent accidental re-use of a free'd pooled instance.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs [hunk 8]
```diff
@@ -51,7 +51,7 @@ internal class AsyncMethodToStateMachineRewriter : MethodToStateMachineRewriter
         /// of rewritten return expressions. The return-handling code then uses <c>SetResult</c> on the async method builder
         /// to make the result available to the caller.
         /// </summary>
-        private readonly LocalSymbol _exprRetValue;
+        protected readonly LocalSymbol _exprRetValue;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodToStateMachineRewriter.cs]: >protected [](start = 8, length = 9)

Is this change needed?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs (hunk 1 above)
Comment (agocke) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs]: Ah, I see, thanks

**Comment by jcouv** [comment]
FYI @stephentoub Fixed in VS 16.4p3.

**Comment by stephentoub** [comment]
Thanks!
